File: old/gears.c
/************************************************************************************/
/*

Copyright Jason T. Harris 2004. All rights reserved.

Calculate Change Gears for a 8x14 Lathe.

*/
/************************************************************************************/
// Includes

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

/************************************************************************************/
// Gear Configurations

enum
{
    GEAR_CONFIG_0,    // A to B to C
    GEAR_CONFIG_1,    // AB to CD to E
    GEAR_CONFIG_2,    // AB to C to D
    GEAR_CONFIG_3,    // A to BC to D
    GEAR_CONFIG_MAX   // must be last
};

/************************************************************************************/
// Geometric Constants

#define MM_PER_INCH    25.4
#define PI             3.1415926535
#define IDLER_TEETH    40.0
#define LEADSCREW_TPI  12.0
#define GEAR_PITCH     0.125

/************************************************************************************/
// Geometric Constraints

#define MINIMUM_GTL            154  // Minimum gear train length
#define MIN_FINAL_GEARS        80   // Minimum center to center for final 2 gears
#define MAX_AB_DIFFERENCE      6    // Maximum AB difference for gear configs 1 & 2
#define MIN_GEAR_CLEARANCE     4    // Minimum gear to gear clearance
#define GEAR_X_SIZE            30   // Size of a hub with no gear present
#define MAX_SLOT_GEARS         168  // Maximum length for engaged gears in the slot

/************************************************************************************/
// Gear Set

static int gear_set[] =
{
    30, 35, 40, 45, 48, 50, 60, 66,
    68, 70, 70, 72, 75, 80, 90, 100
};

#define NUM_GEARS (sizeof(gear_set) / sizeof(int))

/************************************************************************************/
// Gear Solutions

#define INITIAL_ERROR 1.0

typedef struct
{
    double error;

    int a;
    int b;
    int c;
    int d;
    int e;

} GS;

static GS soln[ GEAR_CONFIG_MAX ];

/************************************************************************************/
// Variables

static double goal_pitch;
static int valid_count = 0;
static int vec[ 10 ];

/************************************************************************************/
// Return the screw pitch in inches

static double Calc_Pitch( int cfg, int ga, int gb, int gc, int gd, int ge )
{
    double ratio;

    ratio = IDLER_TEETH / (double)ga;

    switch( cfg )
    {
        case GEAR_CONFIG_0:
        {
            ratio *= ( (double)ga / (double)gc );
            break;
        }

        case GEAR_CONFIG_1:
        {
            ratio *= ( (double)gb / (double)gc );
            ratio *= ( (double)gd / (double)ge );
            break;
        }

        case GEAR_CONFIG_2:
        {
            ratio *= ( (double)gb / (double)gd );
            break;
        }

        case GEAR_CONFIG_3:
        {
            ratio *= ( (double)ga / (double)gb );
            ratio *= ( (double)gc / (double)gd );
            break;
        }

        default:
        {
            assert( 0 );
        }
    }

    return ratio / LEADSCREW_TPI;
}

/************************************************************************************/
// Return the gear train length in teeth

static int Calc_Length( int cfg, int ga, int gb, int gc, int gd, int ge )
{
    int length;

    length = IDLER_TEETH / 2;

    switch( cfg )
    {
        case GEAR_CONFIG_0:
        {
            length += (double)ga;
            length += (double)gb;
            length += (double)gc / 2;
            break;
        }

        case GEAR_CONFIG_1:
        {
            length += (double)ga / 2;
            length += (double)gb / 2;
            length += (double)gc / 2;
            length += (double)gd / 2;
            length += (double)ge / 2;
            break;
        }

        case GEAR_CONFIG_2:
        {
            length += (double)ga / 2;
            length += (double)gb / 2;
            length += (double)gc;
            length += (double)gd / 2;
            break;
        }

        case GEAR_CONFIG_3:
        {
            length += (double)ga;
            length += (double)gb / 2;
            length += (double)gc / 2;
            length += (double)gd / 2;
            break;
        }

        default:
        {
            assert( 0 );
        }
    }

    return length;
}

/************************************************************************************/

static void Display_Gears(int cfg, int mode)
{
    int ga, gb, gc, gd, ge;
    double pitch, error;

    error = ( soln[ cfg ].error / goal_pitch ) * 100.0;

    ga = soln[cfg].a;
    gb = soln[cfg].b;
    gc = soln[cfg].c;
    gd = soln[cfg].d;
    ge = soln[cfg].e;

    pitch = Calc_Pitch(cfg, ga, gb, gc, gd, ge);

    if (mode != 0) {
        printf( "desired tpi = %f\n", ( 1.0 / goal_pitch ) );
        printf( "actual tpi = %f\n", ( 1.0 / pitch ) );
    } else {
        printf( "desired pitch = %f mm\n", goal_pitch * MM_PER_INCH );
        printf( "actual pitch = %f mm\n", pitch * MM_PER_INCH );
    }

    printf( "pitch error = %.2f%%\n", error );

    switch (cfg) {
        case GEAR_CONFIG_0: {
            printf( "%d-%d-%d\n", ga, gb, gc );
            break;
        }
        case GEAR_CONFIG_1: {
            printf( "%d:%d-%d:%d-%d\n", ga, gb, gc, gd, ge );
            break;
        }
        case GEAR_CONFIG_2: {
            printf( "%d:%d-%d-%d\n", ga, gb, gc, gd );
            break;
        }
        case GEAR_CONFIG_3: {
            printf( "%d-%d:%d-%d\n", ga, gb, gc, gd );
            break;
        }
        default: {
            assert( 0 );
        }
    }

    printf( "\n" );
}

/************************************************************************************/

static void Display_Gears_HTML( int cfg, int mode )
{
    int ga, gb, gc, gd, ge;
    double pitch, error;

    error = ( soln[ cfg ].error / goal_pitch ) * 100.0;

    ga = soln[ cfg ].a;
    gb = soln[ cfg ].b;
    gc = soln[ cfg ].c;
    gd = soln[ cfg ].d;
    ge = soln[ cfg ].e;

    pitch = Calc_Pitch( cfg, ga, gb, gc, gd, ge );

    printf( "<tr>" );

    if( mode != 0 )
    {
        printf( "<td>%.2f</td>", ( 1.0 / goal_pitch ) );
        printf( "<td>%.3f</td>", ( 1.0 / pitch ) );
    }
    else
    {
        printf( "<td>%.2f</td>", goal_pitch * MM_PER_INCH );
        printf( "<td>%.3f</td>", pitch * MM_PER_INCH );
    }

    printf( "<td>%.2f%%</td>", error );

    printf( "<td>" );

    switch( cfg )
    {
        case GEAR_CONFIG_0:
        {
            printf( "%d-%d-%d", ga, gb, gc );
            break;
        }

        case GEAR_CONFIG_1:
        {
            printf( "%d:%d-%d:%d-%d", ga, gb, gc, gd, ge );
            break;
        }

        case GEAR_CONFIG_2:
        {
            printf( "%d:%d-%d-%d", ga, gb, gc, gd );
            break;
        }

        case GEAR_CONFIG_3:
        {
            printf( "%d-%d:%d-%d", ga, gb, gc, gd );
            break;
        }

        default:
        {
            assert( 0 );
        }
    }

    printf( "</td></tr>\n" );
}

/************************************************************************************/
// Is this a geometrically valid gear configuration?

static int Gear_Valid( int cfg, int ga, int gb, int gc, int gd, int ge )
{
    switch( cfg ) {
        case GEAR_CONFIG_0: {
            if ((ga + gb) > MAX_SLOT_GEARS) {
                return 0;
            }
            if( ( gb + gc ) < MIN_FINAL_GEARS) {
                return 0;
            }
            break;
        }

        case GEAR_CONFIG_1: {
            if( ( gb + gc ) > MAX_SLOT_GEARS ) {
                return 0;
            }

            if( ( gd + ge ) < MIN_FINAL_GEARS ) {
                return 0;
            }

            if ( ( gb - ga ) > MAX_AB_DIFFERENCE ) {
                return 0;
            }

            if( ( gb + gc ) - ( ga + gd ) < MIN_GEAR_CLEARANCE ) {
                return 0;
            }

            if( ( gd + ge ) - ( gc + GEAR_X_SIZE ) < MIN_GEAR_CLEARANCE ) {
                return 0;
            }

            break;
        }

        case GEAR_CONFIG_2:
        {
            if( ( gb + gc ) > MAX_SLOT_GEARS )
            {
                return 0;
            }

            if( ( gc + gd ) < MIN_FINAL_GEARS )
            {
                return 0;
            }

            if( ( gb - ga ) > MAX_AB_DIFFERENCE )
            {
                return 0;
            }

            break;
        }

        case GEAR_CONFIG_3:
        {
            if( ( ga + gb ) > MAX_SLOT_GEARS )
            {
                return 0;
            }

            if( ( gc + gd ) < MIN_FINAL_GEARS )
            {
                return 0;
            }

            if( ( ga + gb ) - ( gc + GEAR_X_SIZE ) < MIN_GEAR_CLEARANCE )
            {
                return 0;
            }

            if( ( gc + gd ) - ( gb + GEAR_X_SIZE ) < MIN_GEAR_CLEARANCE )
            {
                return 0;
            }

            break;
        }

        default:
        {
            assert( 0 );
        }
    }

    if( Calc_Length( cfg, ga, gb, gc, gd, ge ) < MINIMUM_GTL )
    {
        return 0;
    }

    return 1;
}

/************************************************************************************/

static void Check_Gears( int cfg )
{
    int ga, gb, gc, gd, ge;

    ga = gear_set[ vec [ 0 ] ];
    gb = gear_set[ vec [ 1 ] ];
    gc = gear_set[ vec [ 2 ] ];
    gd = gear_set[ vec [ 3 ] ];
    ge = gear_set[ vec [ 4 ] ];

    if( Gear_Valid( cfg, ga, gb, gc, gd, ge ) != 0 )
    {
        double error, pitch;

        valid_count ++;

        pitch = Calc_Pitch( cfg, ga, gb, gc, gd, ge );
        error = fabs( pitch - goal_pitch );

        if( error < soln[ cfg ].error )
        {
            soln[ cfg ].error = error;
            soln[ cfg ].a = ga;
            soln[ cfg ].b = gb;
            soln[ cfg ].c = gc;
            soln[ cfg ].d = gd;
            soln[ cfg ].e = ge;
        }
    }
}

/************************************************************************************/

static void swap( int i, int j )
{
    int tmp = vec[ i ];
    vec[ i ] = vec[ j ];
    vec[ j ] = tmp;
}

/************************************************************************************/

static void Permute_Gears( int cfg, int n )
{
    int i;

    if( n == 0 )
    {
        Check_Gears( cfg );
        return;
    }

    for( i = 0; i <= n; i ++ )
    {
        swap( n, i );
        Permute_Gears( cfg, n - 1 );
        swap( n, i );
    }
}

/************************************************************************************/

static void Combine_Gears( int ix, int kx,  int n, int k )
{
    int i;

    if( kx == 3 )
    {
        Permute_Gears( GEAR_CONFIG_0, 2 );
    }

    if( kx == 4 )
    {
        Permute_Gears( GEAR_CONFIG_2, 3 );
        Permute_Gears( GEAR_CONFIG_3, 3 );
    }

    if( kx == k )
    {
        Permute_Gears( GEAR_CONFIG_1, 4 );
        return;
    }

    for( i = ix; i <= n; i ++ )
    {
        vec[ kx ] = i;
        Combine_Gears( i + 1, kx + 1, n, k );
    }
}

/************************************************************************************/

void Search_Gears(int mode)
{
    int i, best;
    double error = INITIAL_ERROR;

    for( i = 0; i < GEAR_CONFIG_MAX; i ++ )
    {
        soln[ i ].error = INITIAL_ERROR;
    }

    Combine_Gears( 0, 0, NUM_GEARS - 1, 5 );

    // Find the best solution

    for( i = 0; i < GEAR_CONFIG_MAX; i ++ )
    {
        if( soln[ i ].error < error )
        {
            error = soln[ i ].error;
            best = i;
        }
    }

    //Display_Gears_HTML( best, mode );
    Display_Gears(best, mode);

    //printf( "Valid Configurations %d\n", valid_count );
}

/************************************************************************************/

int main( int argc, char **argv )
{
    int tpi, mm;

#if 0

    // Inch Sizes
    for (tpi = 7; tpi <= 410; tpi ++) {
        goal_pitch = (1.0 / ((double)tpi / 2.0));
        Search_Gears(1);
    }

    // Metric Sizes
    for (mm = 1; mm <= 75; mm ++) {
        goal_pitch = ((double)mm / 10.0) / MM_PER_INCH;
        Search_Gears(0);
    }

#endif

    goal_pitch = (double)(1.75) / MM_PER_INCH;
    Search_Gears(0);

    exit(0);
}

/************************************************************************************/

File: gears.py
#! /usr/bin/python
#------------------------------------------------------------------------------

import math
import itertools
import html

#------------------------------------------------------------------------------
# common thread pitches

_common_metric = (
    0.20, 0.25, 0.35, 0.40, 0.45, 0.50, 0.60, 0.70,
    0.75, 0.80, 1.00, 1.25, 1.50, 1.75, 2.00, 2.50,
    3.00, 3.50, 4.00, 4.50, 5.00, 5.50, 6.00,
)

_common_inch = (
    8.0, 9.0, 10.0, 11.0, 11.5, 12.0, 13.0, 14.0, 16.0,
    18.0, 20.0, 24.0, 28.0, 32.0, 36.0, 40.0, 44.0,
    48.0, 56.0, 64.0, 72.0, 80.0
)

#------------------------------------------------------------------------------
# geometric constants

_MM_PER_INCH = 25.4
_IDLER_TEETH = 40.0
_LEADSCREW_TPI = 12.0
_GEAR_PITCH = 0.125

#------------------------------------------------------------------------------

# gear configurations
# ab -> gears a and b are on the same shaft
# a-b -> gears a and b are meshed
_gear_cfg = (
    'a-b-c',
    'ab-cd-e',
    'ab-c-d',
    'a-bc-d',
)

# the change gear set (by tooth count)
_gear_set = (
    30.0, 35.0, 40.0, 45.0, 48.0, 50.0, 60.0, 66.0,
    68.0, 70.0, 70.0, 72.0, 75.0, 80.0, 90.0, 100.0
)

#------------------------------------------------------------------------------

def calc_pitch(cfg, gears):
    """return the screw pitch in inches"""

    (a, b, c, d, e) = gears

    ratio = _IDLER_TEETH / a

    if cfg == 'a-b-c':
        ratio *= (a / c)
    elif cfg == 'ab-cd-e':
        ratio *= (b / c)
        ratio *= (d / e)
    elif cfg == 'ab-c-d':
        ratio *= (b / d)
    elif cfg == 'a-bc-d':
        ratio *= (a / b)
        ratio *= (c / d)
    else:
        assert(0)

    return ratio / _LEADSCREW_TPI

#------------------------------------------------------------------------------

def calc_length(cfg, gears):
    """return the gear train length in teeth"""

    (a, b, c, d, e) = gears

    n = _IDLER_TEETH / 2.0

    if cfg == 'a-b-c':
        n += (a + b + (c / 2.0))
    elif cfg == 'ab-cd-e':
        n += ((a / 2.0) + (b / 2.0) + (c / 2.0) + (d / 2.0) + (e / 2.0))
    elif cfg == 'ab-c-d':
        n += ((a / 2.0)+ (b / 2.0) + c + (d / 2.0))
    elif cfg == 'a-bc-d':
        n += (a + (b / 2.0) + (c / 2.0) + (d / 2.0))
    else:
        assert(0)

    return n

#------------------------------------------------------------------------------
# geometric constraints

_MINIMUM_GTL = 154.0        # Minimum gear train length
_MIN_FINAL_GEARS = 80.0     # Minimum center to center for final 2 gears
_MAX_AB_DIFFERENCE = 6.0    # Maximum AB difference for gear configs 1 & 2
_MIN_GEAR_CLEARANCE = 4.0   # Minimum gear to gear clearance
_GEAR_X_SIZE = 30.0         # Size of a hub with no gear present
_MAX_SLOT_GEARS = 168.0     # Maximum length for engaged gears in the slot

def is_valid(cfg , gears):
    """return True if this is a geometrically valid gear configuration"""

    (a, b, c, d, e) = gears

    if cfg == 'a-b-c':
        if (a + b) > _MAX_SLOT_GEARS:
            return False
        if (b + c) < _MIN_FINAL_GEARS:
            return False
    elif cfg == 'ab-cd-e':
        if (b + c) > _MAX_SLOT_GEARS:
            return False
        if (d + e) < _MIN_FINAL_GEARS:
            return False
        if (b - a) > _MAX_AB_DIFFERENCE:
            return False
        if (b + c) - (a + d) < _MIN_GEAR_CLEARANCE:
            return False
        if (d + e) - (c + _GEAR_X_SIZE) < _MIN_GEAR_CLEARANCE:
            return False
    elif cfg == 'ab-c-d':
        if (b + c) > _MAX_SLOT_GEARS:
            return False
        if (c + d) < _MIN_FINAL_GEARS:
            return False
        if (b - a) > _MAX_AB_DIFFERENCE:
            return False
    elif cfg == 'a-bc-d':
        if (a + b) > _MAX_SLOT_GEARS:
            return False
        if (c + d) < _MIN_FINAL_GEARS:
            return False
        if (a + b) - (c + _GEAR_X_SIZE) < _MIN_GEAR_CLEARANCE:
            return False
        if (c + d) - (b + _GEAR_X_SIZE) < _MIN_GEAR_CLEARANCE:
            return False
    else:
        assert(0)

    if calc_length(cfg, gears) < _MINIMUM_GTL:
        return False

    return True

#------------------------------------------------------------------------------

def generate_sets(sets, cfg, gear_set):
    """generate pitch solutions for a given gear configuration and set of gears"""
    for gears in gear_set:
        if is_valid(cfg, gears):
            sets.append((calc_pitch(cfg, gears), cfg, gears))

def generate_change_gears():
    """generate a sorted list of change gear sets across all gear configurations and gear sets"""
    sets = []
    # generate the gear permutations
    _3gears = tuple(itertools.permutations(_gear_set, 3))
    _4gears = tuple(itertools.permutations(_gear_set, 4))
    _5gears = tuple(itertools.permutations(_gear_set, 5))
    # pad all gear sets to 5 elements with dummys
    _3gears = tuple([(a, b, c, 0, 0)for (a, b, c) in _3gears])
    _4gears = tuple([(a, b, c, d, 0)for (a, b, c, d) in _4gears])
    # generate the solutions for each gear configuration
    generate_sets(sets, 'a-b-c', _3gears)
    generate_sets(sets, 'ab-c-d', _4gears)
    generate_sets(sets, 'a-bc-d', _4gears)
    generate_sets(sets, 'ab-cd-e', _5gears)
    sets.sort()
    return sets

#------------------------------------------------------------------------------

def search_gears(sets, pitch):
    """return the list index of the gear configuration most closely matching the pitch"""
    if not sets:
        return None
    # do a binary search of the sorted solution list
    hi = len(sets) - 1
    lo = 0
    while hi - lo > 1:
        i = (hi + lo) / 2
        if sets[i][0] >= pitch:
            # move to lower pitch values
            hi = i
        else:
            # move to higher pitch values
            lo = i
    # pick the index with the lowest error
    hi_error = sets[hi][0] - pitch
    lo_error = pitch - sets[lo][0]
    return (hi, lo)[lo_error <= hi_error]

#------------------------------------------------------------------------------

def gears_txt(soln, goal, mode):

    s = []
    (pitch, cfg, gears) = soln

    if mode == 'inch':
        s.append('desired tpi = %f' % (1.0 / goal))
        s.append('actual tpi = %f' % (1.0 / pitch))
    elif mode == 'metric':
        s.append('desired pitch = %f mm' % (goal * _MM_PER_INCH))
        s.append('actual pitch = %f mm' % (pitch * _MM_PER_INCH))
    else:
        assert(0)

    error = math.fabs(pitch - goal)
    error = (error / goal) * 100.0
    s.append('pitch error = %.2f%%' % error)

    (a, b, c, d, e) = gears

    if cfg == 'a-b-c':
        s.append('%d-%d-%d' % (a, b, c))
    elif cfg == 'ab-cd-e':
        s.append('%d:%d-%d:%d-%d' % (a, b, c, d, e))
    elif cfg == 'ab-c-d':
        s.append('%d:%d-%d-%d' % (a, b, c, d))
    elif cfg == 'a-bc-d':
        s.append('%d-%d:%d-%d' % (a, b, c, d))
    else:
        assert(0)

    return '\n'.join(s)


def gears_html(soln, goal, mode):

    td = []
    (pitch, cfg, gears) = soln

    if mode == 'inch':
        td.append(html.tag_td('%.1f' % (1.0 / goal)))
        td.append(html.tag_td('%.2f' % (1.0 / pitch)))
    elif mode == 'metric':
        td.append(html.tag_td('%.2f' % (goal * _MM_PER_INCH)))
        td.append(html.tag_td('%.3f' % (pitch * _MM_PER_INCH)))
    else:
        assert(0)

    error = math.fabs(pitch - goal)
    error = (error / goal) * 100.0
    td.append(html.tag_td('%.2f%%' % error))

    (a, b, c, d, e) = gears

    if cfg == 'a-b-c':
        td.append(html.tag_td('%d-%d-%d' % (a, b, c)))
    elif cfg == 'ab-cd-e':
        td.append(html.tag_td('%d:%d-%d:%d-%d' % (a, b, c, d, e)))
    elif cfg == 'ab-c-d':
        td.append(html.tag_td('%d:%d-%d-%d' % (a, b, c, d)))
    elif cfg == 'a-bc-d':
        td.append(html.tag_td('%d-%d:%d-%d' % (a, b, c, d)))
    else:
        assert(0)

    return html.indent('\n'.join(td))

#------------------------------------------------------------------------------

_ofile = 'stuff.html'

def main():

    # generate the complete set of valid change gears
    sets = generate_change_gears()

    inch_solns = []
    for tpi in _common_inch:
        pitch = 1.0 / float(tpi)
        inch_solns.append((search_gears(sets, pitch), pitch))

    metric_solns = []
    for mm in _common_metric:
        pitch = float(mm) / _MM_PER_INCH
        metric_solns.append((search_gears(sets, pitch), pitch))

    if _ofile.endswith('txt'):
        s = []
        for (i, pitch) in inch_solns:
            s.append(gears_txt(sets[i], pitch, 'inch'))
        for (i, pitch) in metric_solns:
            s.append(gears_txt(sets[i], pitch, 'metric'))
        f = open(_ofile, 'w')
        f.write('\n'.join(s))
        f.close()


    if _ofile.endswith('html'):
        page = []
        page.append(html.include('info.txt'))

        page.append((html.tag_h2('Inch Solutions')))
        th = []
        th.append(html.tag_th('Desired Pitch (tpi)'))
        th.append(html.tag_th('Actual Pitch (tpi)'))
        th.append(html.tag_th('Pitch Error'))
        th.append(html.tag_th('Configuration'))
        tr = []
        tr.append(html.tag_tr(html.indent('\n'.join(th))))
        for (i, pitch) in inch_solns:
            tr.append(html.tag_tr(gears_html(sets[i], pitch, 'inch')))
        table = html.tag_table(html.indent('\n'.join(tr)), 'border="1" cellpadding="5"')
        page.append(table)

        page.append((html.tag_h2('Metric Solutions')))
        th = []
        th.append(html.tag_th('Desired Pitch (mm)'))
        th.append(html.tag_th('Actual Pitch (mm)'))
        th.append(html.tag_th('Pitch Error'))
        th.append(html.tag_th('Configuration'))
        tr = []
        tr.append(html.tag_tr(html.indent('\n'.join(th))))
        for (i, pitch) in metric_solns:
            tr.append(html.tag_tr(gears_html(sets[i], pitch, 'metric')))
        table = html.tag_table(html.indent('\n'.join(tr)), 'border="1" cellpadding="5"')
        page.append(table)

        # build the body
        body = []
        # add the content
        body.append('\n'.join(page))
        body = html.tag_body(html.indent('\n'.join(body)))
        # build the head
        head = []
        head = html.tag_head(html.indent('\n'.join(head)))

        f = open(_ofile, 'w')
        f.write(html.tag_html(html.indent('\n'.join((head, body)))))
        f.close()


#------------------------------------------------------------------------------

main()

#------------------------------------------------------------------------------
